Modbus controller

martgras opened the pull request:
# What does this implement/fix? 

Implements a generic modbus controller to control modbus slaves. 
This PR replaces  #1256 with a generic implementation that isn't limit to EPEVER. 
#1256 required modifications to the exiting 'modbus' component while this version only needs UART.  The modbus specifics are included in the code. The existing 'modbus' component is limited  to input registers and adding support to write data to the slave required a fair amount of changes.   

The implementation has been tested with a couple of modbus slaves including EPEVER Solar Charger, Heidelberg Wallbox (implemented by @derwolff)  , PZEM-017.   The example folder contains these definitions. 

@nikito7 implemented support for the EDP BOX that seems to be popular in Portugal (see https://github.com/nikito7/edp_box_modbus/blob/dev/esphome/edpbox1.yaml  ) 

Sensors, Binary Sensors, Text Sensors and Switches are implemented. 
Writing non boolean data to the slave can be done from a lambda see the epever.yaml  for an example.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1388
  
# Test Environment

- [x] ESP32
- [x] ESP8266
- [ ] Windows
- [ ] Mac OS
- [ ] Linux

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
esphome:
  name: solarstation
  platform: ESP32
  board: esp32dev

substitutions:
  updates: 30s

wifi:
  ssid: !secret wifi_sid
  password: !secret wifi_password
  reboot_timeout: 5min

time:
  - platform: sntp
    id: sntp_time
    timezone: "CET-1CEST,M3.5.0,M10.5.0/3"
    servers: pool.ntp.org

uart:
  id: mod_bus
  tx_pin: 17
  rx_pin: 16
  baud_rate: 115200
  stop_bits: 1


modbus:
  flow_control_pin: 5
  send_wait_time: 200ms
  id: mod_bus_epever

modbus_controller:
  - id: epever
    address: 0x1
    modbus_id: mod_bus_epever
    command_throttle: 0ms
    setup_priority: -10
    update_interval: 30s

sensor:
  - platform: modbus_controller
    modbus_controller_id: epever
    id: array_rated_voltage
    name: "array_rated_voltage"
    address: 0x3000
    unit_of_measurement: "V"
    register_type: read
    value_type: U_WORD
    accuracy_decimals: 1
    skip_updates: 0
    filters:
      - multiply: 0.01

  - platform: modbus_controller
    modbus_controller_id: epever
    id: array_rated_current
    name: "array_rated_current"
    address: 0x3001
    unit_of_measurement: "A"
    register_type: read
    value_type: U_WORD
    accuracy_decimals: 2
    filters:
      - multiply: 0.01

  - platform: modbus_controller
    modbus_controller_id: epever
    id: array_rated_power
    name: "array_rated_power"
    address: 0x3002
    unit_of_measurement: "W"
    register_type: read
    value_type: U_DWORD_R
    accuracy_decimals: 1
    filters:
      - multiply: 0.01

binary_sensor:
  - platform: modbus_controller
    modbus_controller_id: epever
    id: charging_input_volt_failure
    name: "Charging Input Volt Failure"
    register_type: read
    address: 0x3201
    bitmask: 0xC000

switch:
  - platform: modbus_controller
    modbus_controller_id: epever
    id: manual_control_load
    register_type: coil
    address: 2
    name: "manual control the load"
    bitmask: 1

  - platform: modbus_controller
    modbus_controller_id: epever
    id: default_control_the_load
    register_type: coil
    address: 3
    name: "default control the load"
    bitmask: 1


text_sensor:
  - platform: modbus_controller
    modbus_controller_id: epever
    name: "rtc clock test "
    id: rtc_clock_test2
    internal: true
    register_type: holding
    address: 0x9013
    register_count: 3
    raw_encode: HEXBYTES
    response_size: 6

```

# Explain your changes

There are a couple of feature requests for general modbus support. e.g. https://github.com/esphome/feature-requests/issues/952 
https://github.com/esphome/feature-requests/issues/670

This is a new implementation allows a declarative implementation for modbus slaves using RTU. Actually support the PZEM-0xx devices that use modbus can be implemented without writing more code

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).
Working on it : https://github.com/martgras/esphome-docs .

nikito7 reviewed (changes requested):
Run ```script/build_codeowners.py```

nikito7 reviewed (changes requested):
Edit: my mistake

nikito7 commented:
@glmnet can you check this? Regards

syssi commented on esphome/components/modbus_controller/modbuscontroller.cpp:
```diff
@@ -0,0 +1,427 @@
+}
+
+void ModbusController::dump_config() {
+  ESP_LOGCONFIG(TAG, "EPSOLAR:");
```
```suggestion
  ESP_LOGCONFIG(TAG, "ModbusController:");
```

syssi commented on esphome/components/modbus_controller/modbuscontroller.h:
```diff
@@ -0,0 +1,242 @@
+  READ_HOLDING_REGISTERS = 0x03,
+  READ_INPUT_REGISTERS = 0x04,
+  WRITE_SINGLE_COIL = 0x05,
+  WRITE_SINGLE_REGISTER = 0x6,
```
```suggestion
  WRITE_SINGLE_REGISTER = 0x06,
```

syssi commented on esphome/components/modbus_controller/modbuscontroller.h:
```diff
@@ -0,0 +1,242 @@
+  READ_INPUT_REGISTERS = 0x04,
+  WRITE_SINGLE_COIL = 0x05,
+  WRITE_SINGLE_REGISTER = 0x6,
+  READ_EXCEPTION_STATUS = 0x7,   // not implemented
```
```suggestion
  READ_EXCEPTION_STATUS = 0x07,   // not implemented
```

syssi commented on esphome/components/modbus_controller/modbuscontroller.h:
```diff
@@ -0,0 +1,242 @@
+  WRITE_SINGLE_COIL = 0x05,
+  WRITE_SINGLE_REGISTER = 0x6,
+  READ_EXCEPTION_STATUS = 0x7,   // not implemented
+  DIAGNOSTICS = 0x8,             // not implemented
```
```suggestion
  DIAGNOSTICS = 0x08,             // not implemented
```

syssi commented on esphome/components/modbus_controller/modbuscontroller.cpp:
```diff
@@ -0,0 +1,427 @@
+#include "esphome/components/mqtt/mqtt_component.h"
+#include "esphome/components/mqtt/mqtt_switch.h"
+#endif
+#include "modbuscontroller.h"
```
Did you choose the filename by intention? I would expect `modbus_controller.{cpp,h}` here.

OttoWinter commented on esphome/components/modbus_controller/examples/test.yaml:
```diff
@@ -0,0 +1,138 @@
+esphome:
```
We should not commit these examples, they're huge files that can become outdated quickly

martgras replied:
reverted to use modbus

OttoWinter commented on esphome/components/modbus_controller/__init__.py:
```diff
@@ -0,0 +1,283 @@
+            cv.Optional(
+                CONF_COMMAND_THROTTLE, default="0ms"
+            ): cv.positive_time_period_milliseconds,
+            cv.Optional("sensors"): cv.All(
```
Use constant names here like `CONF_SENSORS` - if they're not in const.py you can define them in this module.

OttoWinter commented on esphome/components/modbus_controller/__init__.py:
```diff
@@ -0,0 +1,283 @@
+                )
+            )
+    if config.get("binary_sensors"):
+        conf = config["binary_sensors"]
```
What is the difference between this and the `binary_sensor` platform you have defined?

martgras replied:
Kind of historic  :) 
The code started with hardcoded registers for an epever device and then I was convinced to turn it into a more generic component.  

There are 2 different ways to define a sensor. Either inline like here (this was the original approach I started with) 

````yaml
modbus_controller:
  - id: modbus_controller_test
    address: 0x2
    modbus_id: mod_bus1
    sensors:
      - id: battery_voltage
        name: "Battery voltage"
        address: 0x331A
        offset: 0
        modbus_functioncode: "read_input_registers"
        value_type: U_WORD
      - id: battery_current
        name: "Battery current"
        address: 0x331B
        offset: 2
        modbus_functioncode: "read_input_registers"
        value_type: U_WORD
````
or using the "traditional" way 

````yaml
switch:
  - platform: modbus_controller
    modbus_controller_id: modbus_controller_test
    id: reset_to_fabric_default
    name: "Reset to Factory Default"
    address: 0x9008
    modbus_functioncode: write_single_coil
````

Both variants generate the same C++ Code . In my opinion the first variant is easier to write if you have a lot registers.
Especially when more than one register is read in single command.  The code detects that there are 2 consecutive addresses and sends a read multiple register command instead 2 read single register commands

jesserockz replied:
I don't think there should be multiple places to define `sensors` etc. I would prefer for the config to be the traditional ESPHome way (second option above only) with the code in subfolders and less `AUTO_LOAD`ing in the main file which will allow for cutting down on binary size, as right now just including the modbus controller will include all of the sensor, binary, text, switch code in various places.

martgras replied:
changed as proposed.

OttoWinter commented on esphome/components/modbus_controller/modbusbase.cpp:
```diff
@@ -0,0 +1,241 @@
+      *p_++ = b; \
+  }
+
+#define DUMP_LOG() \
```
We have the `hexencode` helper method for this

OttoWinter commented on esphome/components/modbus_controller/modbusbase.h:
```diff
@@ -0,0 +1,67 @@
+
+  /** RX,TX Control pin Ref: https://github.com/greays/esphome/blob/master/esphome/components/rs485/rs485.h */
+  void set_ctrl_pin(uint8_t ctrl_pin) {
+    static GPIOPin PIN(ctrl_pin, OUTPUT);
```
No, this should be passed in as a GPIOPin directly, see for example `switch.gpio` platform

OttoWinter commented on esphome/components/modbus_controller/readme.md:
```diff
@@ -0,0 +1,506 @@
+# Modbus controller component
+
+## A fork of [esphome](https://github.com/esphome/esphome) adding support for monitoring a modbus device
```
Stale comment, and we document in the docs repo, not here.

OttoWinter commented on esphome/components/modbus_controller/modbuscontroller.h:
```diff
@@ -0,0 +1,242 @@
+  std::queue<std::unique_ptr<ModbusCommandItem>> incoming_queue_;
+  uint32_t last_command_timestamp_;
+  uint16_t command_throttle_;
+  static std::atomic_bool sending_;  // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
```
ESPHome is single-threaded, so atomics are unneeded (plus bools are atomic on pretty much all architectures anyway for the way it is used here)

martgras replied:
Thanks for the review.   My main motivation  for not using the existing component was making it easier to use the component without having to change the existing one.   My approach here didn't work for more than one modbus device on a single bus and therefore I have kept working on an improved version.   
It's now in a new branch at https://github.com/martgras/esphome/tree/multidev

@syssi  and your suggestions are implemented in that branch . Examples and readme's removed but instead I modified the original modbus component to handle writes and not only reads.  

Implemented all suggested changes and updated PR

OttoWinter reviewed (commented):
Could you elaborate why this can't be put into the existing modbus integration?

I really don't think we should have two integrations implementing the same thing, as that just gets super confusing for users and developers.

The existing `modbus` is only used in a couple integrations anyway, if we need to change the base design we can do that.

martgras committed:
Add support for generic modbus devices
Extend modbus component  with support for more function codes

martgras committed:
Add support for FP32 datatype
Improve  multi device parsing

martgras committed:
add  "send_wait_time" parameter to modbus

martgras committed:
add debug code for FP32

martgras committed:
Merge branch 'esphome:dev' into testing

martgras committed:
Poll switch registers
if  a read function code is used poll the values

hnykda commented:
Just curious - is there anything a random esphome user who wants this can help to get this merged? :eyes:

clomads commented:
Seconding as a random ESPHome user who's looking forward to this for use with an EPEVER Tracer. Anything I can do to help?

jesserockz committed:
Simplify config and generated code

jesserockz commented:
@martgras Please take a look at https://github.com/martgras/esphome/pull/29

martgras committed:
Merge pull request #29 from jesserockz/modbus-controller

Simplify config and generated code

martgras committed:
use device addr for lock

martgras committed:
merge changes from testing

martgras committed:
merge changes from testing 2

martgras committed:
Merge branch 'esphome:dev' into modbus_controller

martgras committed:
fix lint errors

martgras committed:
calculate register_count

martgras committed:
change 'offset' to words (registers)
add 'byte_offset' as an alternative to specifiy offsets in bytes

martgras committed:
Tidy send_with_payload and rename to send

martgras committed:
Rename arg to be less confusing
buffer_offset vs register offset

martgras committed:
Don't return value from parse_and_publish

martgras committed:
Never skip switch updates

martgras committed:
Lint

martgras committed:
Auto combine sequential addresses
(remove the need to work with offsets)
add force_new_range property to disable auto combine

martgras committed:
Switch to codebranch used for PR

martgras committed:
Lint

martgras committed:
Lint

martgras committed:
revert offset = 2 * byte_offset
it breaks existing setups and using offset isn't required for most cases

martgras committed:
Merge branch 'testing' into modbus_controller

martgras committed:
Merge branch 'esphome:dev' into modbus_controller

martgras committed:
update readme

martgras committed:
Merge branch 'esphome:dev' into testing

martgras committed:
Merge branch 'esphome:dev' into modbus_controller

martgras committed:
start adding code documentation.
Add a handler paramater to create_custom_command

martgras committed:
Merge branch 'testing' into modbus_controller

martgras committed:
Merge branch 'modbus_controller' of https://github.com/martgras/esphome into modbus_controller

martgras committed:
fix bug in latest commit (no commands send)

martgras committed:
fix bug in latest commit (no commands send)

martgras committed:
Merge branch 'modbus_controller' of https://github.com/martgras/esphome into modbus_controller

martgras committed:
Add support for modbus number

martgras committed:
add support for modbus output

martgras committed:
Merge branch 'testing' into modbus_controller

martgras committed:
Merge branch 'esphome:dev' into testing

martgras committed:
Lint

martgras committed:
Merge branch 'testing' into modbus_controller

martgras committed:
Lint

Fix lint run errors

martgras committed:
Merge branch 'testing' of https://github.com/martgras/esphome into testing

martgras committed:
refactor data conversation

martgras committed:
Add lambda option
provides access to the raw modbus data

martgras committed:
Merge branch 'testing' into modbus_controller

martgras committed:
breaking change - deprecate modbus_function_code. Switch from modbus_function_code to register type in yaml

martgras committed:
lint

martgras committed:
lint

martgras committed:
Update readme and add item param to lambda

martgras committed:
Merge branch 'testing' into modbus_controller

martgras committed:
Update test

martgras committed:
fix return statement

martgras committed:
Merge branch 'esphome:dev' into modbus_controller

jesserockz commented on esphome/components/modbus/modbus.cpp:
```diff
@@ -49,48 +55,74 @@ bool Modbus::parse_modbus_byte_(uint8_t byte) {
+       return false;
+     }
+   }
+   */
```
Don't leave commented code in please =)
```suggestion
```

martgras committed:
fix modbus switch crash after  after #2303
can't pass 'this' as const to non-const functions

jesserockz commented on esphome/components/modbus_controller/__init__.py:
```diff
@@ -0,0 +1,150 @@
+    return "not found"
+
+
+# Helper method to translate function_code to register_type
+# Warn the user if function_code is still in used
+# Can be removed after some time
+def set_register_type(cfg):
+    if CONF_REGISTER_TYPE in cfg:
+        reg_type = cfg[CONF_REGISTER_TYPE]
+    else:
+        reg_type = function_code_to_register(cfg[CONF_MODBUS_FUNCTIONCODE])
+        reg_type_name = str(reg_type)
+        reg_type_name = reg_type_name[reg_type_name.rfind("::") + 2 :].lower()
+        _LOGGER.warning(
+            "%s has been been deprected. Component id=%00s\n        Change '%s: %s' to '%s: %s' ",
+            CONF_MODBUS_FUNCTIONCODE,
+            cfg[CONF_ID],
+            CONF_MODBUS_FUNCTIONCODE,
+            cfg[CONF_MODBUS_FUNCTIONCODE],
+            CONF_REGISTER_TYPE,
+            reg_type_name,
+        )
+    return reg_type
```
Considering this is a new component, this cannot be deprecated already

martgras replied:
true but there are already a couple of users working with it from my repo. 
https://github.com/martgras/esphome/discussions/11
I will remove it once the PR is ready for merge

martgras replied:
removed

jesserockz commented on esphome/components/modbus_controller/output/modbus_output.h:
```diff
@@ -0,0 +1,46 @@
+
+#include "esphome/components/output/float_output.h"
+#include "esphome/components/modbus_controller/modbus_controller.h"
+#include "esphome/components/modbus_controller/sensor/modbus_sensor.h"
```
This wont work if the user does not have any modbuscontroller sensors as the file will not be copied to the generated project.

Individual platforms should not rely on each other, only the parent component that is in the `DEPENDENCIES`

martgras replied:
fixed

jesserockz commented on esphome/components/modbus_controller/text_sensor/modbus_textsensor.cpp:
```diff
@@ -0,0 +1,59 @@
+  auto result = output.str();
+  // Is there a lambda registered
+  // call it with the pre converted value and the raw data array
+  if (this->transform_func_.has_value()) {
+    // the lambda can parse the response itself
+    auto val = (*this->transform_func_)(this, result, data);
+    if (val.has_value()) {
+      ESP_LOGV(TAG, "Value overwritten by lambda");
+      result = val.value();
+    } else {
+      ESP_LOGV(TAG, "publishing handled by lambda - parse and publish");
+      return;
+    }
+  }
+  this->publish_state(result);
```
I don't think the lambdas should be responsible for publishing the state, they should just transform the value only.

**Same for all types, not just text_sensor**

martgras replied:
I can't find anymore from which component I copied this concept :)
Basically the idea is that a return {} from the lambda means  stop processing after the lambda: "everything has been taken care of" - but  yes maybe not really required here. 
I just followed the pattern from the number lambda  where it makes sense to give you the chance to handle more complex scenarios
I'm a bit undecided: 
-  remove that code path or 
- change the log message something like  " ESP_LOGV(TAG, "Empty return from lambda - no further processing");

jesserockz replied:
Like I said, I would prefer if the lambda was only used for transformation, and publishing was done in the component code. I see no real reason to have a user publish inside the lambda and will end up causing confusion/issues.

martgras replied:
ok will remove it

jesserockz commented on esphome/components/modbus_controller/output/__init__.py:
```diff
@@ -0,0 +1,71 @@
+    "ModbusOutput", cg.Component, output.FloatOutput, SensorItem
+)
+
+CONF_MODBUS_SENSOR_ID = "modbus_sensor_id"
```
Is this used?

martgras replied:
no

jesserockz reviewed (changes requested):
While reading the docs I realised I missed this:

Number should not publish to a Sensor, the Number itself should change to the value read from the controller and should be independent.

martgras commented:
If Numbers had filters that would be my preference as well. 
They are a good fit for read/write registers
But they lack filters so to decouple them from sensor I'd have  at least provide a lambda for publishing the value . Maybe another multiply property...
The best option is adding filters to numbers ( for publish and write to device)

jesserockz commented:
Maybe the `lambda` options should be consistently used for converting the register value to the desired frontend value, and for things like number (and select if you decide to implement that) could have a second lambda `write_lambda` which converts to the register value.

martgras committed:
remove dead code

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

nikito7 commented:
Change link to: https://github.com/nikito7/edpbox/tree/dev/esphome

nikito7 commented:
@martgras you re-added support for 4 bytes registers?

```yaml
  - platform: modbus_controller
    name: "${dev} T1 Vazio"
    address: 0x0026
    offset: 0
    modbus_functioncode: "read_input_registers"
    value_type: U_DWORD
    accuracy_decimals: 1
    filters:
      - multiply: 0.001
    unit_of_measurement: "kWh"
    device_class: energy
    state_class: total_increasing
```

martgras commented:
> you re-added support for 4 bytes registers?
> ```yaml
>  - platform: modbus_controller
>     name: "${dev} T1 Vazio"
>     address: 0x0026
>     offset: 0
>     modbus_functioncode: "read_input_registers"
>     value_type: U_DWORD
>     accuracy_decimals: 1
>     filters:
>       - multiply: 0.001
>     unit_of_measurement: "kWh"
>     device_class: energy
>     state_class: total_increasing
> ```

Yes  but then you have to set register_count: 1.
The code assumes 2 registers for DWORD therefore you have to override it
(Please have a look at https://github.com/martgras/esphome/blob/testing/esphome/components/modbus_controller/readme.md#note-sep15 since there a couple of other changes)

```yaml
 - platform: modbus_controller
    name: "${dev} T1 Vazio"
    address: 0x0026
    register_type: "read"
    value_type: U_DWORD
   register_count: 1
   accuracy_decimals: 1
   filters:
     - multiply: 0.001
   unit_of_measurement: "kWh"
   device_class: energy
   state_class: total_increasing
> ```

martgras commented:
> Maybe the `lambda` options should be consistently used for converting the register value to the desired frontend value, and for things like number (and select if you decide to implement that) could have a second lambda `write_lambda` which converts to the register value.

Technically agreed  but unsure about the naming

For incoming data (filter) I keep the name 'lambda'
For number , output and (in the future select the name) will be 'write_lambda' for  write to device actions.

ok?

martgras committed:
Implement review suggestions#
 - remove empty lambda return path
 - refactor number (remove sensor dependency)

martgras committed:
Merge branch 'esphome:dev' into modbus_controller

martgras committed:
remove register_type from number
holding register is the only writeable register

nikito7 reviewed (approved):
From my needs. All is working

martgras committed:
remove modbus function code.
 remove empty lambda return  from switch.

jesserockz commented on esphome/components/modbus_controller/switch/modbus_switch.cpp:
```diff
@@ -0,0 +1,70 @@
+static const char *const TAG = "modbus_controller.switch";
+
+void ModbusSwitch::setup() {
+  // value isn't required 
```
```suggestion
  // value isn't required
```

jesserockz committed:
Remove trailing whitespace

jesserockz reviewed (approved):
LGTM! Great job =)

reaper7 commented:
how this implementation will affect the SDM energy meter component ?

jesserockz commented:
> how this implementation will affect the SDM energy meter component ?

It does not affect it. This is a new component that allows generic access to any modbus device without building it in native ESPHome code. Native ESPHome components will still provide a minimal/less confusing way for people to use those devices though (like SDM)
